taichi/transforms/scalarize.cpp

```diff
@@ -554,7 +554,23 @@ class ScalarizePointers : public BasicStmtVisitor {
 
         // TODO(zhanlue): loose this contraint once dynamic indexing is properly
         // handled
-        TI_ASSERT(stmt->offset->is<ConstStmt>());
+        if (!stmt->offset->is<ConstStmt>()) {
+          modifier_.modify_ir();
```

Comment:
Shall we get rid of this line since we are immediately raising an exception.

Reply:
Removing this line will trigger an assertion error in the destructor of `modifier_`: https://github.com/taichi-dev/taichi/blob/b6b431764eb48f06a651f9afc60e3e28f8629c12/taichi/ir/ir.cpp#L409-L416

Reply:
Destructors in C++ are marked as `noexcept` by default, thus use `TI_ASSERT` in this case can cause undefined behavior. 

The general idea is that `destructors` should try not get interrupted whatsoever, otherwise there's always a risk of resource leak since it violates RAII (part of the resources release codes won't run if a `throw` occurs)

It might worth a comment here.

Reply:
Comments have been added.